### PR TITLE
MINOR: Client state machine fix for transition to stable on initial empty assignment

### PR DESCRIPTION
**Pull request by lianetm**
This includes a fix for ensuring that a member transitions to STABLE when joining a group, in the case where it receives an initial empty assignment.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java [hunk 1]
```diff
@@ -405,8 +405,8 @@ private void processAssignmentReceived(ConsumerGroupHeartbeatResponseData.Assign
                 assignmentUnresolved, currentAssignment);
             // Make sure we transition the member back to STABLE if it was RECONCILING (ex.
             // member was RECONCILING unresolved assignments that were just removed by the
-            // broker).
-            if (state == MemberState.RECONCILING) {
+            // broker), or JOINING (member joining received empty assignment).
+            if (state == MemberState.RECONCILING || state == MemberState.JOINING) {
                 // This is the case where a member was RECONCILING an unresolved
                 // assignment that was removed by the broker in a following assignment.
```
Comment (cadonna) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: This comment is not completely correct anymore. It only describes one case of the two. Doesn't this comment  duplicate the comment above the `if`-statement.
I really strongly recommend to not use too many inline comments. They are really hard to properly maintain as we can see here.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java [hunk 2]
```diff
@@ -1403,6 +1403,14 @@ public void testHeartbeatSentOnStaledMember() {
         assertEquals(MemberState.JOINING, membershipManager.state());
     }
 
+    @Test
+    public void testMemberJoiningTransitionsToStableWhenReceivingEmptyAssignment() {
```
Comment (cadonna) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java:1407]: Is there also a test for the second case? I skimmed over the test class but could not find one.

**Review by cadonna — COMMENTED** [review]
Thanks for the PR, @lianetm !

Here my feedback!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java (hunk 1 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java]: You're right, the comment is a dup of the one above and was not accurate anymore, removed (agree with the suggestion of not overdoing the inline comments, noted)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java (hunk 2 above)
Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java:1407]: Yes, there is, `testNewEmptyAssignmentReplacesPreviousOneWaitingOnMetadata` covers the other case.

**Comment by lianetm** [comment]
Thanks a lot for the review @cadonna . All comments addressed.

**Review by cadonna — APPROVED** [review]
LGTM!

**Comment by cadonna** [comment]
@lianetm 
Are the following failures related?
```
Build / JDK 11 and Scala 2.13 / kafka.api.PlaintextConsumerTest.testShrinkingTopicSubscriptions(String, String).quorum=kraft+kip848.groupProtocol=consumer
Build / JDK 17 and Scala 2.13 / kafka.api.PlaintextConsumerTest.testExpandingTopicSubscriptions(String, String).quorum=kraft+kip848.groupProtocol=consumer
Build / JDK 17 and Scala 2.13 / kafka.api.PlaintextConsumerTest.testExpandingTopicSubscriptions(String, String).quorum=kraft+kip848.groupProtocol=consumer
```

**Comment by lianetm** [comment]
@cadonna those are unrelated, they had been failing in other PRs as well and are now disabled (disabled with @philipnee latest [PR](https://github.com/apache/kafka/pull/14937#issuecomment-1862111535) that I just merged here along with trunk latest changes). Thanks!

**Comment by lianetm** [comment]
Test failures seem unrelated. I checked around `testCoordinatorFailover` that, even though it passes locally, it failed on the CI for the new consumer, but it's been failing for other recent PRs and reported flaky with [KAFKA-16024](https://issues.apache.org/jira/browse/KAFKA-16024).

**Pull request merged**